man/man1/rocotorewind.1
.\"-
.\" Manual page written by Christopher Harrop <christopher.w.harrop@noaa.gov>
.\"
.\"
.Dd January 25, 2013
.Os
.Dt ROCOTOBOOT 1 CON
.Sh NAME
.Nm rocotorewind
.Nd undo the effect of running specific jobs
.Sh SYNOPSIS
.Nm
.Op Fl h
.Op Fl v Ar #
.Op Fl -version
.Fl d Ar database_file
.Fl w Ar workflow_file
.Fl t Ar task
.Fl c Ar cycle
.Sh DESCRIPTION
.Nm
is a tool that attempts to undo the effect of running jobs.  This
command is most commonly used to rerun part of a workflow that has
failed.  It edits the Rocoto database to remove evidence that the job
is run.  If any <rewind> tags are given for a rewound <task>, then the
<sh> and <rb> actions within are run.  That allows post-job cleanup,
if necessary.  If the -a option (all tasks) is given, then the
database is edited to remove all entries related to the cycle, making
it as if the cycle had never been run.  The following limitations
apply:
.Bl -bullet -offset indent -compact
.It
A task cannot be rewound if its state is EXPIRED.
.It
A task cannot be rewound for a cycle is EXPIRED.
.El
The
.Nm
command requires two arguments. The
.Cm -w
flag specifies the name
of the workflow definition file. The
.Cm -d
flag specifies the name of the database file
that contains the state of the workflow for the task that is to be rewound.
.Pp
Additionally, the rocotorewind will not do anything unless the -c option
and at least one of -a, -m, or -t are specified.
.Pp
The following options are available:
.Bl -tag -width flag
.It Fl -version
Display the version number and exit.
.It Fl h, -help
Display a short help message and exit.
.It Fl v, -verbose Ar #
Run Rocoto in verbose mode with the specified level. If no level is
specified a level of 1 is used.
.It Fl d, -database Ar file
The path to the Rocoto database file for the current workflow.
.It Fl w, -workflow Ar file
The path to the Rocoto workflow file.
.It Fl c, --cycle Ar cyclespec[,cyclespec[,...]]
Only submit jobs for these cycles. See
.Sx WORKFLOW SUBSETTING
.It Fl m, --metatasks Ar metataskspec[,metataskspec[,...]]
Only submit jobs for these metatasks.  See
.Sx WORKFLOW SUBSETTING
.It Fl t, --tasks Ar taskspec[,taskspec[,...]
Only submit jobs for these tasks.  See
.Sx WORKFLOW SUBSETTING
.It Fl a, --all
Include all tasks in the workflow subset.  See
.Sx WORKFLOW SUBSETTING
.El
.Sh WORKFLOW SUBSETTING
All Rocoto commands that monitor or modify the workflow can be told to
run on some subset of that workflow.  Subset specifications use the
-a, -t, -c, and -m options.  Using multiple options adds to the
selection.  The rocotorun program will only submit jobs if they're in
the selected subset.
.Pp
The resulting set is the set union of the -t, -m, and -a option, for
each cycle selected by the -c option.

.Ss All Tasks: The -a Option
The -a option has no arguments, and selects all possible tasks.  It is
an error to specify -t or -m if the -a option is given.

.Ss Task Subset: The -t Option
This consists of one or more comma-separated lists of task
specifications which describe the set of tasks.  Without a leading
dash, every taskspec adds to the set of selected tasks (set union).
With a dash, the taskspec is removed (set subtraction) from the list
of tasks.
.Bl -tag flag
.It Ar -taskspec[,taskspec[,...]]
Negate the option (remove tasks from the selection instead of adding).
This applies to all taskspecs following the "-".
.It Ar /regex/
Include all tasks whose names match the given regular expression
(everything within the two slashes).  Typical regular expression
characters have their usual meanings: ( [ ] ) | . * +
.It Ar :attribute
Select all tasks that have the specified XML attribute set for its
<task> tag, and the attribute is true, non-zero, or non-empty.
Supported attributes are final, shared, exclusive, cores, and nodes.
The special value "metatasks" selects tasks that are in a metatask
(any metatask).
.It Ar @cycledef
Select all tasks in the given cycledef
.It Ar taskname
Select a task with the specified name.
.El

.Ss Metatask Subset: The -m Option
This consists of one or more comma-separated lists of metatask
specifications which describe the set of metatasks.  Without a leading
dash, every metataskspec adds to the set of selected metatasks (set
union).  With a dash, the metataskspec is removed (set subtraction)
from the list of metatasks.
.Bl -tag flag
.It Ar -metataskspec[,metataskspec[,...]]
Negate the option (remove metatasks from the selection instead of adding).
This applies to all metataskspecs following the "-".
.It Ar metataskname
Add a metatask to the selected tasks.
.El

.Ss Cycle subset: The -c Option
Unlike the -t and -m options, the -c option can only include one
cyclespec, rather than a comma-separated list (except individual
cycles; see below).  Every cyclespec adds to the set of selected
cyclespec (set union).  Set subtraction is NOT supported (no -
options).
.Bl -tag flag
.It Ar YYYYMMDDHHMM[,YYYYMMDDHHMM[,...]]
Select specific cycles, given their twelve-digit date/time specification.
.It Ar YYYYMMDDHHMM:YYYYMMDDHHMM
Select the specified range of cycles.
.It Ar :YYYYMMDDHHMM
Select all cycles from the first one to the given cycle.
.It Ar YYYYMMDDHHMM:
Select all cycles from the given cycle to the last known cycle.
.It Ar all | ALL
Select all cycles.
.It @cycledef
Select all cycles in the specified cycledef.
.El

.Sh EXAMPLES
The command:
.Pp
.Ic "rocotorewind -d foo.db -w foo.xml -t wrf -c 201301010000"
.Pp
Will remove from the database
.Qq wrf
task for the cycle
.Qq 201301010000
defined in the workflow definition file named
.Qq foo.xml
and contained in the database file named
.Qq foo.db
If the <task> had any <rewind> tags, the rewind actions (<sh> and
<rb>) within are run.
.Sh EXIT STATUS
.Ex -std
.Sh DOCUMENTATION
Complete documentation can be found at: http://rdhpcs.noaa.gov/rocoto/
.Sh SEE ALSO
.Xr rocotoboot 1 ,
.Xr rocotocomplete 1 ,
.Xr rocotocheck 1 ,
.Xr rocotorun 1 ,
.Xr rocotostat 1
.Sh STANDARDS
The
.Nm
command is expected to be
.St -p1003.2
compatible.
.Sh AUTHOR
Written by Christopher Harrop
.Sh BUGS
Report bugs to <christopher.w.harrop@noaa.gov>


RELEASE_NOTES.md
# Release Notes

## New for Version 1.3.0

* Update SLURM support to handle pack groups
* Update SLURM support to map `<queue>` to --qos instead of --partition
* Update SLURM support to map `<partition>` to --partition
* Update LSF support to handle additional methods LSF uses to report the exit status
* All rocoto commands have the same -a, -c, -m, and -t options
* The -c and -t options can now select by cycledefs and attributes (ie. final)
* The new "rocotocomplete" command can mark tasks or cycles as having completed
* Manpages are updated to reflect current capabilities

## New for Version 1.2.4

* Fix bugs relating to ~/.rocoto/log rotation
* Fix race condition bug relating to ~/.rocoto/rocotorc configuration file
* Increase reliability and performance of database and workflow locking
* Fix bug in Rocoto commands related to sourcing of shell init script
* Improve performance and reduce system load when monitoring for orphaned Rocoto processes
* Fix bug in XML validation that erroneously enforced ordering of metatask contents
* Fix bug in processing of XML special characters
* Fix task list update bug in rocotorewind
* Fix bug in rotocoboot that prevent task output files from being rotated
* Fix bugs in Cobalt batch system interface that were exposed in Ruby > 1.8.7
* Fix bug in LSF batch system interface related to long job names
* Add support for PBSPro batch system (e.g. for use on Cheyenne)
* Update libxml-ruby to version 3.0.0 to mitigate memory management bugs
* Add new cyclestring flags
  * @n = Number of days in the month
  * @o = All lower case abbreviated month name
  * @O = All lower case full month name
* Add new &lt;taskvalid&gt; dependency to trigger tasks based on whether or not a particular task is valid for the current cycle time


sbin/rocotorewind.rb
#!/usr/bin/ruby

# Get the base directory of the WFM installation
__WFMDIR__=File.expand_path("../../",__FILE__)

# Add include paths for WFM and libxml-ruby libraries
$:.unshift("#{__WFMDIR__}/lib")
$:.unshift("#{__WFMDIR__}/lib/rubysl-date/lib")
$:.unshift("#{__WFMDIR__}/lib/rubysl-parsedate/lib")
$:.unshift("#{__WFMDIR__}/lib/libxml-ruby")
$:.unshift("#{__WFMDIR__}/lib/sqlite3-ruby")
$:.unshift("#{__WFMDIR__}/lib/SystemTimer")
$:.unshift("#{__WFMDIR__}/lib/open4/lib")
$:.unshift("#{__WFMDIR__}/lib/thread/lib")

# Load workflow engine library
require 'workflowmgr/workflowengine'
require 'workflowmgr/workflowsubsetoptions'
require 'workflowmgr/utilities'
require 'libxml'

# Replace that ridiculous Libxml-ruby handler that automatically sends
# output to stderr We want to control what output goes where and when.
LibXML::XML::Error.set_handler do |error|
  WorkflowMgr.stderr(error.to_s)
  WorkflowMgr.log(error.to_s)
end

# Create workflow engine and run it
opt=WorkflowMgr::WorkflowSubsetOptions.new(ARGV,'rocotorewind','rewind')
workflowengine=WorkflowMgr::WorkflowEngine.new(opt)
workflowengine.rewind!
